compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnnotationArgumentReference.kt

```diff
@@ -73,43 +123,117 @@ public sealed class AnnotationArgumentReference {
+        }
+
+        return parent?.findConstantDefinitionInScope(name)
+      }
+
+      fun parsePrimitiveType(text: String): Any {
+        // We need to check the parameter for what type exactly.
+        return when (parameterFqName()) {
+          Boolean::class.qualifiedName, BooleanArray::class.qualifiedName ->
+            text.toBooleanStrictOrNull()
+          Int::class.qualifiedName, IntArray::class.qualifiedName ->
+            text.toIntOrNull()
+          Long::class.qualifiedName, LongArray::class.qualifiedName ->
+            text.toLongOrNull()
+          Double::class.qualifiedName, DoubleArray::class.qualifiedName ->
+            text.toDoubleOrNull()
+          Byte::class.qualifiedName, ByteArray::class.qualifiedName ->
+            text.toByteOrNull()
+          Short::class.qualifiedName, ShortArray::class.qualifiedName ->
+            text.toShortOrNull()
+          Float::class.qualifiedName, FloatArray::class.qualifiedName ->
+            text.toFloatOrNull()
+          String::class.qualifiedName -> {
+            if (text.startsWith("\"") && text.endsWith("\"")) {
+              text.substring(1, text.length - 1)
+            } else {
+              text
+            }
+          }
+          else -> fail()
+        } ?: fail()
+      }
+
+      fun parsePsiElement(psiElement: PsiElement): Any {
+        return when (psiElement) {
+          is KtClassLiteralExpression -> psiElement.requireFqName(module).toClassReference(module)
+          is KtCollectionLiteralExpression ->
+            psiElement.children.map { parsePsiElement(it) }.convertToArrayIfNeeded()
+          is KtConstantExpression -> parsePrimitiveType(psiElement.text)
+          is KtStringTemplateExpression ->
+            psiElement
+              .getChildOfType<KtStringTemplateEntry>()
+              ?.text
+              ?: fail()
+          is KtNameReferenceExpression -> {
+            // Those are likely enum values.
+            psiElement.fqNameOrNull(module)?.let { return it }
+
+            // Check if it's a constant. Note that we don't return the constant FqName directly,
+            // but return the actual value from the constant instead. That's also what the
+            // descriptor APIs do. They never see the constant itself, but the
+            // resolved / inlined value instead.
+            psiElement
+              .findConstantDefinitionInScope(psiElement.getReferencedName())
+              ?.let { return parsePrimitiveType(it) }
+
+            // If we reach here, it's a top-level constant defined in the same package but
+            // a different file. In this case, we can just do the same in our generated code
+            // because we're in the same package and can play by the same rules.
+            return psiElement.containingKtFile.packageFqName
+              .descendant(psiElement.getReferencedName())
+          }
+          is KtDotQualifiedExpression -> psiElement.requireFqName(module)
+          else -> fail()
```

Comment:
I had to heavily extend the parsing of annotation arguments. Psi is quite challenging for that. But I'm happy about the result. Descriptors are so much easier.